src/NuGet.Clients/PackageManagement.UI/Feeds/MultiSourcePackageFeed.cs

```diff
@@ -0,0 +1,226 @@
+            if (notCompleted != null)
+            {
+                aggregated.SourceSearchStatus
+                    .AddRange(notCompleted
+                        .ToDictionary(kv => kv.Key, kv => GetLoadingStatus(kv.Value.Status)));
+            }
+
+            return aggregated;
+        }
+
+        private static LoadingStatus GetLoadingStatus(TaskStatus taskStatus)
+        {
+            switch(taskStatus)
+            {
+                case TaskStatus.Canceled:
+                    return LoadingStatus.Cancelled;
+                case TaskStatus.Created:
+                case TaskStatus.RanToCompletion:
+                case TaskStatus.Running:
+                case TaskStatus.WaitingForActivation:
+                case TaskStatus.WaitingForChildrenToComplete:
+                case TaskStatus.WaitingToRun:
+                    return LoadingStatus.Loading;
+                case TaskStatus.Faulted:
+                    return LoadingStatus.ErrorOccured;
+                default:
+                    return LoadingStatus.Unknown;
+            }
+        }
+
+        private async Task<SearchResult<IPackageSearchMetadata>> AggregateSearchResultsAsync(
+            string searchText, 
+            IEnumerable<SearchResult<IPackageSearchMetadata>> results)
+        {
+            SearchResult<IPackageSearchMetadata> result;
+
+            var nonEmptyResults = results.Where(r => r.Any()).ToArray();
+            if (nonEmptyResults.Length == 0)
+            {
+                result = SearchResult.Empty<IPackageSearchMetadata>();
+            }
+            else if (nonEmptyResults.Length == 1)
+            {
+                result = SearchResult.FromItems(nonEmptyResults[0].Items);
+            }
+            else
+            {
+                var indexer = new LuceneSearchResultsIndexer();
+                var aggregator = new SearchResultsAggregator(indexer);
+                var aggregatedItems = await aggregator.AggregateAsync(
+                    searchText, nonEmptyResults.Select(r => r.Items).ToArray());
+                result = SearchResult.FromItems(aggregatedItems.ToArray());
+            }
+
+            result.SourceSearchStatus = results
+                .SelectMany(r => r.SourceSearchStatus)
+                .ToDictionary(kv => kv.Key, kv => kv.Value);
+
+            var cursors = results
+                .Where(r => r.NextToken != null)
+                .ToDictionary(r => r.SourceSearchStatus.Single().Key, r => r.NextToken);
+
+            if (cursors.Keys.Any())
+            {
+                result.NextToken = new AggregatedContinuationToken
+                {
+                    SearchString = searchText,
+                    SourceSearchCursors = cursors
+                };
+            }
+
+            return result;
+        }
+
+        private void LogError(Task task)
+        {
+            foreach (var ex in task.Exception.Flatten().InnerExceptions)
```

Comment:
There are helpers for this and exception utilities, if possible this should use one of the existing ones, or add a utility for this. This sort of thing is done all over the code.